apps/omg_watcher/lib/exit_validator/validator.ex

```diff
@@ -14,24 +14,65 @@
 
 defmodule OMG.Watcher.ExitValidator.Validator do
   @moduledoc """
-  Fragment of imperative shell for ExitValidator. Validates exits.
+    Contains 'challenge_fastly_invalid_exits' and 'challenge_slowly_invalid_exits' functions
+    resposible for validating exits.
+
+    The exit validators covers following cases:
+      a) (FV) The invalid exit is processed after the tx so it's known to be invalid from the start. Causes `:invalid_exit`
+      b) (SV) The exit is processed before invalid tx, but invalid tx is within `M_SV, so the chain is valid.
+         Causes `:invalid_exit`. Doesn't cause an exit.
+      c) (BG) The exit is processed before invalid tx, but invalid tx is after `M_SV, so the chain is invalid.
+         Causes `:invalid_block`.
+
   """
+  use OMG.API.LoggerExt
 
+  alias OMG.API.EventerAPI
   alias OMG.API.Utxo
+  alias OMG.Watcher.Eventer.Event
   require Utxo
 
-  @spec challenge_invalid_exits(fun()) :: ([OMG.API.State.Core.exit_t()] -> :ok)
-  def challenge_invalid_exits(utxo_exists_callback) do
+  @doc """
+    Validates exits and if exit is invalid then emits `:invalid_exit` event
+  """
+  @spec challenge_fastly_invalid_exits() :: ([OMG.API.State.Core.exit_t()] -> :ok)
+  def challenge_fastly_invalid_exits do
+    challenge_invalid_exits(fn utxo_exit ->
+      if not OMG.API.State.utxo_exists?(utxo_exit) do
+        emit_invalid_exit_event(utxo_exit)
+      end
+    end)
+  end
+
+  @doc """
+    Validates and spends exits in the "OMG.API.State" and if exit is invalid then emits `:invalid_exit` event
+  """
+  @spec challenge_slowly_invalid_exits() :: (fun() -> :ok)
+  def challenge_slowly_invalid_exits do
+    challenge_invalid_exits(fn utxo_exit ->
+      with :ok <- OMG.API.State.exit_if_not_spent(utxo_exit) do
+        _ = Logger.info(fn -> "Spent exit: #{inspect(utxo_exit)}" end)
+      else
+        :utxo_does_not_exist ->
+          emit_invalid_exit_event(utxo_exit)
+
+        other ->
+          other
+      end
+    end)
+  end
+
+  defp challenge_invalid_exits(utxo_exists_callback) do
```

Comment:
this looks like an `Enum.map`. I also find it really hard to understand what's going on here, due to the callbacks being used. Can we simplify? The behavior is relatively simple: 
```
for every exit, 
   if invalid, event
   if valid and going slowly, spend in State
```
more or less. You also could do the work with `_ = OMG.Watcher.DB.EthEvent.insert_exits(exits)` from here, rather than from `Application` - `Application` should in fact be minimal, with no code at all - just configuration

Reply:
yes, thanks 
`  defp challenge_invalid_exits(is_slow_validator) do
    fn utxo_exits ->
      for utxo_exit <- utxo_exits do
        cond do
          not OMG.API.State.utxo_exists?(utxo_exit) ->
            EventerAPI.emit_events([struct(Event.InvalidExit, utxo_exit)])

          not is_slow_validator ->
            _ = OMG.Watcher.DB.EthEvent.insert_exits([utxo_exit])

          is_slow_validator ->
            :ok = OMG.API.State.exit_utxos([utxo_exit])
            _ = Logger.info(fn -> "Spent exit: #{inspect(utxo_exit)}" end)
        end
      end

      :ok
    end
  end`

Reply:
yes, thanks 
```
  defp challenge_invalid_exits(is_slow_validator) do
    fn utxo_exits ->
      for utxo_exit <- utxo_exits do
        cond do
          not OMG.API.State.utxo_exists?(utxo_exit) ->
            EventerAPI.emit_events([struct(Event.InvalidExit, utxo_exit)])

          not is_slow_validator ->
            _ = OMG.Watcher.DB.EthEvent.insert_exits([utxo_exit])

          is_slow_validator ->
            :ok = OMG.API.State.exit_utxos([utxo_exit])
            _ = Logger.info(fn -> "Spent exit: #{inspect(utxo_exit)}" end)
        end
      end

      :ok
    end
  end
```

Reply:
yes, thanks a lot !
```
 defp challenge_invalid_exits(is_slow_validator) do
    fn utxo_exits ->
      for utxo_exit <- utxo_exits do
        cond do
          not OMG.API.State.utxo_exists?(utxo_exit) ->
            EventerAPI.emit_events([struct(Event.InvalidExit, utxo_exit)])

          not is_slow_validator ->
            _ = OMG.Watcher.DB.EthEvent.insert_exits([utxo_exit])

          is_slow_validator ->
            :ok = OMG.API.State.exit_utxos([utxo_exit])
            _ = Logger.info(fn -> "Spent exit: #{inspect(utxo_exit)}" end)
        end
      end

      :ok
    end
  end
```